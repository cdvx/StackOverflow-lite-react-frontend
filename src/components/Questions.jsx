import React, { Component } from 'react';
import { connect } from "react-redux";
import questions from "../actions/questionsAction";

export const Question = ({question}) => {
    return (  
        <React.Fragment >
            <blockquote id="blockQ">
                <p className="text-black" ><strong>Topic: </strong>{question.topic}</p>
                <p id="text-body1">{question.body}</p>
                <small  className="text-black"><strong>Author: </strong>{question.author}</small>
            </blockquote>
        </React.Fragment>
    );
}


const Info = ({messages}) => {
    return (  
        <React.Fragment>
            <tr className="info">
            <td>message</td>
            <td>{messages}</td>
            </tr>
        </React.Fragment>
    );
}

class Questions extends Component {
    state = {  
        questions: [],
        messages: ""
    };

    componentWillMount(){
        const {getQuestions} = this.props;
        getQuestions();
    }

    handleQuestionClick =  questionId => {               
        localStorage.setItem("questionId", questionId);
    }

    renderQuestions = questions => (questions.map(question=>{
        return (
        <React.Fragment>
                <tr >
                <td id="quest"><a id="select-question" href="/question" onClick={()=>(this.handleQuestionClick(question.questionId))}><Question question={question} /></a></td>
                </tr>
                <div id="space-this"></div>

        </React.Fragment>
        )
    }));
    render() { 
        const {results, messages} = this.props;
        console.log("props here>>", this.props, results, messages)
        return ( 
            <React.Fragment>
                <table className="table table-striped table-hover ">
                    <thead>
                        <tr>
                        <div className="container"><th><big >Questions</big></th></div>
                        </tr>
                    </thead>
                    <tbody>
                        {results && this.renderQuestions(results)}
                        {messages && <Info messages={messages}/>}
                    
                    </tbody>
                </table> 
            </React.Fragment>
         );
    }
}
 

export const mapStateToProps = ({questions}) => ({
    results: questions.questions,
    messages: questions.messages
});

export const mapDispatchToProps = dispatch => ({
    getQuestions: () => dispatch(questions()),
    getQuestion: questionId => dispatch(question(questionId))
});

export default connect(mapStateToProps, mapDispatchToProps)(Questions);