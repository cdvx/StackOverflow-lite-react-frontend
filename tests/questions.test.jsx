import { mount, shallow } from "enzyme";

import React from "react";
import expect from "expect";
import { MemoryRouter } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import Questions, { Info, Question, mapStateToProps, mapDispatchToProps} from "../src/components/Questions";


const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({ questionsReducer: {questions: [],messages: "rrff"}, getQuestions: jest.fn()});

it("it should render component", () => {
  shallow(
    <MemoryRouter>
      <Provider store={store}><Questions /></Provider>
    </MemoryRouter>,
  );
});

const props = {
  queston: {
    topic: "User with this name or password not found !",
    body:"ggggggg",
    author: "rfghhj"
  },
  messages: "ffggggg",
  onClick: jest.fn(),
  loginUser: jest.fn(),
  followedUsername: "cdvx",
};

describe("<Questions />", () => {
  test("renders the component", () => {
    const wrapper = shallow(<Provider store={store}><Questions /></Provider>);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("<Question />", () => {
    test("renders the component", () => {
      const wrapper = shallow(<Question question={props.messages}/>);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("<Info />", () => {
    test("renders the component", () => {
      const wrapper = shallow(<Info messages={props.question}/>);
      expect(wrapper).toMatchSnapshot();
    });
  });



describe("test the login container", () => {
  const fakeEvent = { preventDefault: () => {}};

  const wrapper = mount(<Provider store={store}><Questions fakeEvent /></Provider>);
  wrapper.setState({
    questons: [],
    messages: "cdvfnfjgjgjgj"
  });

  wrapper.setProps({loginUser: jest.fn(), onChangeField: jest.fn(fakeEvent, "username") });
  it("should mount the login component", () => {

    expect(wrapper.find("Info")).toBeTruthy();
    expect(wrapper.find("Question")).toBeTruthy();
    // expect(wrapper.find("input").last().simulate("change", fakeEvent));

  });
});

describe("Login", () => {
  it("should show previously previously passed state", () => {
    const state = {
        "questionsReducer": [{topic:"xxx",body:"FFFF", author:"Vddcd"}]
    };
    expect(mapStateToProps(state)).toEqual({"questionsReducer": [{topic:"xxx",body:"FFFF", author:"Vddcd"}]});

  });

  it("Should pass action to props on submit", () => {
    const dispatch = jest.fn();
    const fakeEvent = { preventDefault: () => {}};
    mapDispatchToProps(dispatch).getQuestions();
    expect((dispatch.mock.calls[0][0])).toBeTruthy();
  });
});


describe("Checks whether mapstate to props returns", () => {
  const expectedProp = {
    state:{ state: {username: "",password: ""},getQuestions: jest.fn()},
    getQuestions: jest.fn()
  };

  expect(mapStateToProps(expectedProp)).toBeTruthy();
});


describe("Checks whether mapdispatch to props returns", () => {
  const dispatch = jest.fn();
  mapDispatchToProps(dispatch).getQuestions();
  expect(dispatch.mock.calls.length).toBe(1);
});
