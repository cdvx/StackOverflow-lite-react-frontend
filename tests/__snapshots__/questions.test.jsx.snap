// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Info /> renders the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Info />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <tr
        className="info"
      >
        <td>
          message
        </td>
        <td />
      </tr>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <td>
            message
          </td>,
          <td />,
        ],
        "className": "info",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "message",
          },
          "ref": null,
          "rendered": "message",
          "type": "td",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "td",
        },
      ],
      "type": "tr",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <tr
          className="info"
        >
          <td>
            message
          </td>
          <td />
        </tr>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <td>
              message
            </td>,
            <td />,
          ],
          "className": "info",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "message",
            },
            "ref": null,
            "rendered": "message",
            "type": "td",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "td",
          },
        ],
        "type": "tr",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<Question /> renders the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Question
    question="ffggggg"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <blockquote
        id="blockQ"
      >
        <p
          className="text-black"
        >
          <strong>
            Topic: 
          </strong>
        </p>
        <p
          id="text-body1"
        />
        <small
          className="text-black"
        >
          <strong>
            Author: 
          </strong>
        </small>
      </blockquote>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="text-black"
          >
            <strong>
              Topic: 
            </strong>
          </p>,
          <p
            id="text-body1"
          />,
          <small
            className="text-black"
          >
            <strong>
              Author: 
            </strong>
          </small>,
        ],
        "id": "blockQ",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <strong>
                Topic: 
              </strong>,
              undefined,
            ],
            "className": "text-black",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Topic: ",
              },
              "ref": null,
              "rendered": "Topic: ",
              "type": "strong",
            },
            undefined,
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "id": "text-body1",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <strong>
                Author: 
              </strong>,
              undefined,
            ],
            "className": "text-black",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Author: ",
              },
              "ref": null,
              "rendered": "Author: ",
              "type": "strong",
            },
            undefined,
          ],
          "type": "small",
        },
      ],
      "type": "blockquote",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <blockquote
          id="blockQ"
        >
          <p
            className="text-black"
          >
            <strong>
              Topic: 
            </strong>
          </p>
          <p
            id="text-body1"
          />
          <small
            className="text-black"
          >
            <strong>
              Author: 
            </strong>
          </small>
        </blockquote>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p
              className="text-black"
            >
              <strong>
                Topic: 
              </strong>
            </p>,
            <p
              id="text-body1"
            />,
            <small
              className="text-black"
            >
              <strong>
                Author: 
              </strong>
            </small>,
          ],
          "id": "blockQ",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <strong>
                  Topic: 
                </strong>,
                undefined,
              ],
              "className": "text-black",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Topic: ",
                },
                "ref": null,
                "rendered": "Topic: ",
                "type": "strong",
              },
              undefined,
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "id": "text-body1",
            },
            "ref": null,
            "rendered": null,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <strong>
                  Author: 
                </strong>,
                undefined,
              ],
              "className": "text-black",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Author: ",
                },
                "ref": null,
                "rendered": "Author: ",
                "type": "strong",
              },
              undefined,
            ],
            "type": "small",
          },
        ],
        "type": "blockquote",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<Questions /> renders the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Connect(Questions) />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Connect(Questions) />,
      "value": Object {
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeState": Object {
          "getQuestions": [MockFunction],
          "questionsReducer": Object {
            "messages": "rrff",
            "questions": Array [],
          },
        },
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": Object {},
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Connect(Questions) />,
        "value": Object {
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "storeState": Object {
            "getQuestions": [MockFunction],
            "questionsReducer": Object {
              "messages": "rrff",
              "questions": Array [],
            },
          },
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
